### PR TITLE
Middleware 1.12

### PR DESCRIPTION
**Pull request by IceTank**
Let's try this again without all the commits and force pushes.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -18,14 +18,15 @@
   "types": "lib/index.d.ts",
   "scripts": {
     "build": "npx tsc",
-    "prepublish": "npm run build"
+    "prepublish": "npm run build",
+    "postinstall": "npm run build"
```
Comment (rob9315) [inline package.json]: this shouldn't be needed i think, prepublish already does this iirc

File: src/conn.ts [hunk 2]
```diff
@@ -1,183 +1,376 @@
...
-  //* a whitelist for the internal bot
-  internalWhitelist: string[] = ['keep_alive'];
-  //* filter for the main client as long as it is not overwritten by the client itself
-  toServerBlackList: string[] = ['keep_alive'];
-  //* filter for all attached clients if it is not overwritten by the client
-  toClientBlackList: string[] = ['keep_alive'];
+  optimizePacketWrite: boolean = true;
+  //* Middleware to control packets being send to the client and server
+  toClientMiddleware?: PacketMiddleware[] = [() => {}];
+  //* Middleware to control packets being send to the client and server
```
Comment (rob9315) [inline src/conn.ts]:
```suggestion
  //* Middleware to control packets being sent from the server to the client
  toClientMiddleware?: PacketMiddleware[] = [() => {}];
  //* Middleware to control packets being sent from the client to the server
```

File: src/conn.ts [hunk 3]
```diff
@@ -1,183 +1,376 @@
...
+      bound: 'server' | 'client';
+      /** Only 'packet' is implemented */
+      writeType: 'packet' | 'rawPacket' | 'channel';
+      /** Packet meta. Contains the packet name under `name` also see {@link PacketMeta} */
+      meta: PacketMeta;
+    },
+    /** The client connected to this packet */ pclient: Client,
+    /** Parsed packet data as returned by nmp */ data: any,
+    /** A handle to cancel a packet from being send. Can also force a packet to be send */ cancel: PacketCanceler,
+    /** Indicate that a packet has been modified */ update: packetUpdater
```
Comment (rob9315) [inline src/conn.ts]: I think the cancel and update api is still a bit lacklustre, i would prefer it being controlled be the return type of the function. (I guess this is the chaining kind of Middleware again?)
```suggestion
    // The client connected to this packet
    pclient: Client,
    // Parsed packet data as returned by nmp
    data: any,
    // A handle to cancel a packet from being send. Can also force a packet to be send
    cancel: PacketCanceler,
    // Indicate that a packet has been modified
    update: packetUpdater
```

File: src/conn.ts [hunk 4]
```diff
@@ -1,183 +1,376 @@
...
   bot: Bot & { recipes: number[] };
-  pclient: Client | undefined;
-  pclients: Client[] = [];
+  /** Internal whitelist for the bot */
+  private internalWhitelist: string[] = ['keep_alive'];
+  optimizePacketWrite: boolean = true;
+  /** Contains the currently writing client or undefined if there is none */
+  writingClient: Client | undefined;
+  /** Contains clients that are actively receiving packets from the proxy bot */
+  receivingClients: Client[] = [];
```
Comment (rob9315) [inline src/conn.ts]: pclients is still mentioned in the readme but replaced by these new variables. if possible please change it back to using `pclients`

File: src/conn.ts [hunk 5]
```diff
@@ -1,183 +1,376 @@
...
+          this.bot.entity.onGround = data.onGround;
+          this.bot.emit('move', this.bot.entity.position) // If bot is not in control physics are turned off
+          break;
+        case 'held_item_slot':
+          this.bot.quickBarSlot = data.slot;
+          this.bot._client.emit('mcproxy:heldItemSlotUpdate') // lol idk how to do it better
           break;
         case 'abilities':
-          this.bot.physicsEnabled = !!((data.flags & 0b10) ^ 0b10);
+          this.bot.physicsEnabled = !this.writingClient && !!((data.flags & 0b10) ^ 0b10);
```
Comment (rob9315) [inline src/conn.ts]: why this change?

**Review by rob9315 — CHANGES_REQUESTED** [review]
initial review is in, some minor naming changes please as pointed out in the comments, to realign the code with the readme again.
The way cancelling and update checking works in the middlewares i find to be a bit weird, surely there is a better way

File: src/conn.ts (hunk 5 above)
Comment (IceTank) [reply src/conn.ts]: To disable the bots physics when the bot is not in control to save on performance and prevent the bot from having weird side effects while not being in control off the connection

File: src/conn.ts (hunk 4 above)
Comment (IceTank) [reply src/conn.ts]: Ok

File: src/conn.ts (hunk 2 above)
Comment (IceTank) [reply src/conn.ts]: Fixed

File: package.json (hunk 1 above)
Comment (IceTank) [reply package.json]: Ok

File: src/conn.ts [hunk 6]
```diff
@@ -189,33 +183,45 @@ export class Conn {
...
-      if (!update.isUpdated && this.optimizePacketWrite) {
-        this.writeRaw(buffer);
-        return;
+      if (isCanceled) return
+      if (!wasChanged && this.optimizePacketWrite) {
+        pclient.writeRaw(buffer)
+        return
       }
-      this.write(meta.name, data);
+      pclient.write(meta.name, packetData);
```
Comment (rob9315) [inline src/conn.ts]: i don't think this was supposed to be changed

Comment (rob9315) [reply src/conn.ts]: at least iirc those two should have quite the different meaning

Comment (IceTank) [reply src/conn.ts]: Fixed

File: examples/simpleMiddleware.js [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+});
+
+conn.bot.once('spawn', () => {
+  console.log('spawn');
+
+  /** @type {import('../lib/index').PacketMiddleware} */
+  const filterChatMiddleware = ({ isCanceled, meta }) => {
+    if (isCanceled) return; // Not necessary but may improve performance when using multiple middleware's after each other
+    if (meta.name !== 'chat') return;
+    if (JSON.stringify(data.message).includes('censor')) return false; // Cancel all packets that have the word censor in the chat message string
```
Comment (rob9315) [inline examples/simpleMiddleware.js]: why json.stringify? isn't it already a string of json data?

File: src/conn.ts [hunk 8]
```diff
@@ -1,183 +1,419 @@
...
+    }
+
+    //* keep mineflayer info up to date
+    switch (meta.name) {
+      case 'abilities':
+        let packetData = getPacketData()
+        this.stateData.bot.physicsEnabled = !this.pclient && !!((packetData.flags & 0b10) ^ 0b10);
+      default: // Fallthrough
+        this.stateData.onSToCPacket(meta.name, getPacketData)
+    }
```
Comment (rob9315) [inline src/conn.ts]: idk if this still works correctly, fallthrough should be above the case which doesn't break, not the one after

File: src/conn.ts [hunk 9]
```diff
@@ -1,183 +1,419 @@
...
+      if (this.pclient !== pclient) {
+        return false;
+      }
+      // Keep the bot updated
+      // Note: Packets seam to be the exact same going from server to client and the other way around.
+      // At least for 1.12.2. So this is just copy past from onServerRaw
+      switch (meta.name) {
         case 'abilities':
-          this.bot.physicsEnabled = !!((data.flags & 0b10) ^ 0b10);
+          this.stateData.bot.physicsEnabled = !this.pclient && !!((data.flags & 0b10) ^ 0b10);
```
Comment (rob9315) [inline src/conn.ts]: the bot probably doesn't get set to the correct physics state if this packet was received and ignored unless this info is also stored elsewhere

Comment (rob9315) [reply src/conn.ts]: also, is this not just duplicated code? i think abilities is only sent in one direction, from server to client

File: src/stateData.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+        this.bot.entity.yaw = ((180 - packetData.yaw) * Math.PI) / 180;
+        this.bot.entity.pitch = -(packetData.pitch * Math.PI) / 180;
+        this.bot.entity.onGround = packetData.onGround;
+        break;
+      case 'held_item_slot':
+        packetData = packetDataGetter()
+        this.bot.quickBarSlot = packetData.slot; // Carefull S->C is slot where C->S is slotId (????)
+        break;
+    }
+  }
```
Comment (rob9315) [inline src/stateData.ts]: duplicated code?

**Review by rob9315 — CHANGES_REQUESTED** [review]

File: examples/simpleMiddleware.js (hunk 7 above)
Comment (IceTank) [reply examples/simpleMiddleware.js]: I guess so yes

File: src/conn.ts (hunk 9 above)
Comment (IceTank) [reply src/conn.ts]: Abilities is send both ways. The client can switch from walking to flying with the abilities packet.

File: src/stateData.ts (hunk 10 above)
Comment (IceTank) [reply src/stateData.ts]: Fixed

File: src/conn.ts (hunk 8 above)
Comment (IceTank) [reply src/conn.ts]: This should be fixed now

File: examples/simpleMiddleware.js (hunk 7 above)
Comment (IceTank) [reply examples/simpleMiddleware.js]: Fixed

File: src/conn.ts (hunk 3 above)
Comment (IceTank) [reply src/conn.ts]: Changed

File: src/conn.ts [hunk 11]
```diff
@@ -1,183 +1,410 @@
...
+      { isUpdated: false }
+    );
+
+    // @ts-ignore
+    const funcReturn = this.botToServerDefaultMiddleware({ bound: 'client', meta: { state, name }, writeType: 'packet' }, null, data, cancel, update);
+    if (funcReturn instanceof Promise) {
+      await funcReturn;
+    }
+
+    if (cancel.isCanceled !== false) return;
```
Comment (rob9315) [inline src/conn.ts]: doesn't seem to match the readme?

**Review by rob9315 — COMMENTED** [review]
great, please format the code using the prettier config then it's perfect
